### PR TITLE
Relayer

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Part of #4

## Changes

_Describe the changes that were needed.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium_contracts/deploy-scripts/Cargo.toml [hunk 1]
```diff
@@ -19,10 +19,6 @@ thiserror = "1.0"
 tokio = {version = "1.18", features = ["rt", "macros"]}
 tonic = "0.8"
 
-[dependencies.concordium-contracts-common]
-path = "./deps/concordium-rust-sdk/concordium-base/concordium-contracts-common/concordium-contracts-common"
-version = "4"
-
 [dependencies.concordium-rust-sdk]
-path = "./deps/concordium-rust-sdk"
```
Comment (abizjak) [inline concordium_contracts/deploy-scripts/Cargo.toml:27]: Ignore this. That's just merge artifact.

File: concordium_contracts/deploy-scripts/src/contracts.rs [hunk 2]
```diff
@@ -1,15 +1,10 @@
-use concordium_contracts_common::{Address, ModuleReference, OwnedEntrypointName, OwnedParameter};
```
Comment (abizjak) [inline concordium_contracts/deploy-scripts/src/contracts.rs:1]: Ignore this file.

File: concordium_contracts/deploy-scripts/src/deployer.rs [hunk 3]
```diff
@@ -1,12 +1,12 @@
-use std::str::FromStr;
```
Comment (abizjak) [inline concordium_contracts/deploy-scripts/src/deployer.rs:1]: Ignore this

File: concordium_contracts/deploy-scripts/src/main.rs [hunk 4]
```diff
@@ -8,9 +8,9 @@ use std::{
```
Comment (abizjak) [inline concordium_contracts/deploy-scripts/src/main.rs:8]: Ignore htis.

File: relayer/src/state_sender.rs [hunk 5]
```diff
@@ -0,0 +1,1114 @@
+pub use state_sender::*;
```
Comment (abizjak) [inline relayer/src/state_sender.rs:1]: This is auto-generated. Do not review.

File: relayer/src/root_chain_manager.rs [hunk 6]
```diff
@@ -0,0 +1,1430 @@
+pub use bridge_manager::*;
```
Comment (abizjak) [inline relayer/src/root_chain_manager.rs:1]: This is auto-generated.

File: relayer/src/erc20.rs [hunk 7]
```diff
@@ -0,0 +1,686 @@
+pub use erc_20::*;
```
Comment (abizjak) [inline relayer/src/erc20.rs:1]: This is auto-generated.

**Comment by abizjak** [comment]
@limemloh 

The PR is entirely complete, there is some documentation missing (and some extra add-ons for monitoring, but those will be in the separate PR).

Please look at only the relayer changes. The rest will be resolved once other pending PRs are merged and is not part of this PR.

File: relayer/README.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
+## Relayer
+
+The relayer is the component of the bridge that monitors both the Ethereum and
```
Comment (abizjak) [inline relayer/README.md:3]: Start review here.

File: relayer/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+        v2::Client::new(ep).await?
+    };
+    {
+        let lfb = concordium_client
+            .get_consensus_info()
+            .await?
+            .last_finalized_block;
+        let bi = concordium_client.get_block_info(lfb).await?.response;
+        if chrono::Utc::now().signed_duration_since(bi.block_slot_time)
+            > chrono::Duration::seconds(120)
```
Comment (limemloh) [inline relayer/src/main.rs:312]: Why 120 seconds and not `app.max_behind`?

File: relayer/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+
+    let (start_number, upper_number) = find_start_ethereum_config(
+        ethereum_client.clone(),
+        last_ethereum,
+        app.state_sender_creation_block_number,
+        app.num_confirmations,
+    )
+    .await?;
+    log::info!(
+        "Found starting point at start {}, end {})",
```
Comment (limemloh) [inline relayer/src/main.rs:348]: Consider adding more context to this log. Like this is related to Ethereum

File: relayer/src/concordium_contracts.rs [hunk 11]
```diff
@@ -0,0 +1,685 @@
...
+        match listen_concordium_worker(
+            &mut bridge_manager,
+            &sender,
+            &mut height,
+            max_parallel,
+            max_behind,
+        )
+        .await
+        {
+            Ok(()) => {
```
Comment (limemloh) [inline relayer/src/concordium_contracts.rs:468]: I expected this case to break out of the loop, since this only happens when the database channel is dropped.
Maybe I am missing something

File: relayer/src/concordium_contracts.rs [hunk 12]
```diff
@@ -0,0 +1,685 @@
...
+    let timeout = std::time::Duration::from_secs(max_behind.into());
+    loop {
+        let (error, chunk) = finalized_blocks
+            .next_chunk_timeout(max_parallel as usize, timeout)
+            .await
+            .map_err(|_| NodeError::Timeout)?;
+        let mut futures = futures::stream::FuturesOrdered::new();
+        for fb in chunk {
+            let mut node = bridge_manager.client.clone();
+            futures.push_back(async move {
```
Comment (limemloh) [inline relayer/src/concordium_contracts.rs:535]: Minor thing: Consider triggering poll on `futures` after calling `push_back`. Currently, they are not polled until the while let expression below

File: relayer/src/db.rs [hunk 13]
```diff
@@ -0,0 +1,1272 @@
...
+        }
+        Ok(result)
+    }
+
+    /// Get the transaction hash, the data, and the timestamp when
+    /// the transaction was inserted to the database, in case
+    /// a pending transaction exists.
+    pub async fn pending_ethereum_tx(
+        &self,
+    ) -> anyhow::Result<Option<(H256, ethers::prelude::Bytes, u64, [u8; 32], Vec<u64>)>> {
```
Comment (limemloh) [inline relayer/src/db.rs:508]: Please introduce a struct for the return value.

File: relayer/src/concordium_contracts.rs [hunk 14]
```diff
@@ -0,0 +1,687 @@
...
+                log::info!("Terminated listening for new Concordium events.")
+            }
+            Err(e) => match e {
+                NodeError::Timeout => {
+                    retry_attempt += 1;
+                    if retry_attempt > 6 {
+                        log::error!("Too many failures attempting to reconnect. Aborting.");
+                        anyhow::bail!("Too many failures attempting to reconnect. Aborting.");
+                    }
+                    let delay = std::time::Duration::from_secs(5 << retry_attempt);
```
Comment (limemloh) [inline relayer/src/concordium_contracts.rs:479]: This seems to be missing the actual timeout

File: relayer/src/concordium_contracts.rs [hunk 15]
```diff
@@ -0,0 +1,687 @@
...
+                        delay.as_secs()
+                    );
+                }
+                NodeError::QueryError(e) => {
+                    retry_attempt += 1;
+                    if retry_attempt > 6 {
+                        log::error!("Too many failures attempting to reconnect. Aborting.");
+                        anyhow::bail!("Too many failures attempting to reconnect. Aborting.");
+                    }
+                    let delay = std::time::Duration::from_secs(5 << retry_attempt);
```
Comment (limemloh) [inline relayer/src/concordium_contracts.rs:491]: also here, no timeout?

File: relayer/src/concordium_contracts.rs [hunk 16]
```diff
@@ -0,0 +1,687 @@
...
+    Timeout,
+    /// Query error.
+    #[error("Error querying the node {0}.")]
+    QueryError(#[from] v2::QueryError),
+    /// Internal error. This is a configuration issue.
+    #[error("Internal error: {0}.")]
+    Internal(anyhow::Error),
+}
+
+pub async fn listen_concordium(
```
Comment (limemloh) [inline relayer/src/concordium_contracts.rs:445]: A comment on the retry logic: If I understand this correctly, it will retry 6 times if it fails (unless internal error), but only reset the retry counter when the database connection is closed.
Which seems a bit weird to me.

Node error such as timeouts could accumulate over time and the counter should probably be reset at some point, like every time data is retrieved successfully

Comment (limemloh) [reply relayer/src/concordium_contracts.rs:445]: Maybe this is what you intended to do, but the way `listen_concordium_worker` is defined does not accommodate this. Since it does its own loop.

File: relayer/src/merkle.rs [hunk 17]
```diff
@@ -0,0 +1,735 @@
...
+///   that are to be approved in the database as "tentatively approved".
+/// - After that it waits until the transaction is confirmed on the chain. When
+///   this is done, i.e., the transaction is confirmed, it uses the provided
+///   `db_sender` channel to notify the database to mark the
+///
+/// The response is `Ok(())` if the service was asked to stop, otherwise it is
+/// one of the errors if some part of the job was interrupted.
+/// The client and `pending` are always left in a consistent state, so that a
+/// retry can be made.
+async fn ethereum_tx_sender_worker<M: Middleware, S: Signer>(
```
Comment (limemloh) [inline relayer/src/merkle.rs:507]: Way too much is happening in this function and it is really hard to get through.
Please consider splitting this up a bit

File: relayer/src/main.rs (hunk 9 above)
Comment (abizjak) [reply relayer/src/main.rs:312]: No reason other than I added this as an ad-hoc measure for testing. Will use max_behind instead.

File: relayer/src/main.rs (hunk 10 above)
Comment (abizjak) [reply relayer/src/main.rs:348]: Done.

File: relayer/src/concordium_contracts.rs (hunk 12 above)
Comment (abizjak) [reply relayer/src/concordium_contracts.rs:535]: I don't quite understand what I should do here. I am doing here the same as the ::collect method does.

There is no function `poll` as I see on `FuturesOrdered`. I can see that they refer to it in the documentation, but it is written confusingly. 

I think you're suggesting that I start executing the futures that are pushed immediately. That would make sense, but I don't see how I can do that with the API provided.

In any case I don't think it would affect things much, since the loop has a few iterations only, and the main thing in it is to clone an arc and allocate a future. So I will leave this as is.

File: relayer/src/db.rs (hunk 13 above)
Comment (abizjak) [reply relayer/src/db.rs:508]: Indeed, it grew :)

It is done now.

File: relayer/src/concordium_contracts.rs [hunk 18]
```diff
@@ -0,0 +1,687 @@
...
+    // Flag to signal stopping the task gracefully.
+    mut stop: tokio::sync::watch::Receiver<()>,
+) -> anyhow::Result<()> {
+    // Process the response.
+    // Return an error if submitting this transaction failed and this cannot be
+    // recovered.
+    //
+    // Otherwise either return Ok(true) if retry should be attempted, or Ok(false).
+    // if submission succeeded.
+    let process_response = |hash, response: v2::RPCResult<TransactionHash>| match response {
```
[the remark this answers is not in the document]
Comment (abizjak) [reply relayer/src/concordium_contracts.rs:624]: Because the code is used twice. Once for the normal call and then in retries.

It could have probably been written in a different way ...

File: relayer/src/concordium_contracts.rs [hunk 19]
```diff
@@ -0,0 +1,687 @@
...
+        let retry = process_response(hash, client.send_block_item(&bi).await)?;
+        if retry {
+            // Retry at most 5 times, waiting at most 32 * 5 = 160s
+            let mut success = false;
+            for i in 0..6 {
+                // if the stop sender has been dropped we should terminate since that
+                // should be the last thing that happens in the program.
+                if stop.has_changed().unwrap_or(true) {
+                    log::info!("The service is requesting to terminate. Not retrying.");
+                    return Ok(());
```
[the remark this answers is not in the document]
Comment (abizjak) [reply relayer/src/concordium_contracts.rs:668]: Changed.

File: relayer/src/ethereum.rs [hunk 20]
```diff
@@ -0,0 +1,423 @@
...
+                    anyhow::bail!(
+                        "Too many failures attempting to query Ethereum events. Aborting."
+                    );
+                } else {
+                    let delay = std::time::Duration::from_secs(5 << retry_num);
+                    log::error!(
+                        "Failed getting Ethereum block events due to {e}. Retrying in {} seconds.",
+                        delay.as_secs()
+                    );
+                    retry_num += 1;
```
[the remark this answers is not in the document]
Comment (abizjak) [reply relayer/src/ethereum.rs:199]: Indeed.

File: relayer/src/concordium_contracts.rs (hunk 16 above)
Comment (abizjak) [reply relayer/src/concordium_contracts.rs:445]: Yeah, this is wrong.

Comment (abizjak) [reply relayer/src/concordium_contracts.rs:445]: I have changed it now to reset the retry counter on progress.

File: relayer/src/concordium_contracts.rs (hunk 14 above)
Comment (abizjak) [reply relayer/src/concordium_contracts.rs:479]: Indeed.

Comment (abizjak) [reply relayer/src/concordium_contracts.rs:479]: Fixed now.

File: relayer/src/concordium_contracts.rs (hunk 15 above)
Comment (abizjak) [reply relayer/src/concordium_contracts.rs:491]: Fxied.

File: relayer/src/concordium_contracts.rs (hunk 11 above)
Comment (abizjak) [reply relayer/src/concordium_contracts.rs:468]: No, you are not missing anything.

File: relayer/src/merkle.rs (hunk 17 above)
Comment (abizjak) [reply relayer/src/merkle.rs:516]: The two "wait for pending" and "send" have been split into functions now.

File: relayer/src/merkle.rs [hunk 21]
```diff
@@ -0,0 +1,769 @@
...
+                        // Transaction is confirmed, update the nonce for the next iteration
+                        // of sending.
+                        client.next_nonce += 1.into();
+                        *pending = None;
+                    } else {
+                        log::warn!(
+                            "Ethereum transaction {pending_hash:#x} is in block {bn}, but not yet \
+                             confirmed."
+                        );
+                        *pending = Some((*pending_hash, *root, std::mem::take(ids)));
```
Comment (limemloh) [inline relayer/src/merkle.rs:689]: Small detail: I don't think the assignment is needed here

File: relayer/src/merkle.rs [hunk 22]
```diff
@@ -0,0 +1,769 @@
...
+                    }
+                } else {
+                    return Err(EthereumSenderError::Internal(anyhow::anyhow!(
+                        "A submitted transaction is confirmed, but without a block hash. This \
+                         indicates a configuration error."
+                    )));
+                }
+            } else {
+                log::debug!("Ethereum transaction {pending_hash:#x} is pending.");
+                *pending = Some((*pending_hash, *root, std::mem::take(ids)));
```
Comment (limemloh) [inline relayer/src/merkle.rs:699]: Small detail: I don't think the assignment is needed here

File: relayer/src/merkle.rs [hunk 23]
```diff
@@ -0,0 +1,769 @@
...
+    db_sender: &tokio::sync::mpsc::Sender<DatabaseOperation>,
+    pending: &mut Option<(H256, [u8; 32], Vec<u64>)>,
+    num_confirmations: u64,
+    stop: &mut tokio::sync::watch::Receiver<()>,
+) -> Result<bool, EthereumSenderError<M>>
+where
+    M::Error: 'static,
+    S::Error: 'static, {
+    // Handle followup for any pending transaction first.
+    if pending.is_some() {
```
Comment (limemloh) [inline relayer/src/merkle.rs:594]: I would make this a guard

File: relayer/src/merkle.rs [hunk 24]
```diff
@@ -0,0 +1,769 @@
...
+                return Ok(true);
+            }
+            let result = client
+                .root_manager
+                .client()
+                .get_transaction_receipt(*pending_hash)
+                .await
+                .map_err(EthereumSenderError::Retryable)?;
+            if let Some(receipt) = result {
+                if let Some(bn) = receipt.block_number {
```
Comment (limemloh) [inline relayer/src/merkle.rs:612]: I would make these guards

File: relayer/src/merkle.rs [hunk 25]
```diff
@@ -0,0 +1,769 @@
...
+                    let current_block = client
+                        .root_manager
+                        .client()
+                        .get_block_number()
+                        .await
+                        .map_err(EthereumSenderError::Retryable)?;
+                    if bn.saturating_add(num_confirmations.into()) <= current_block.into() {
```
Comment (limemloh) [inline relayer/src/merkle.rs:619]: I would make this a guard

File: relayer/src/merkle.rs [hunk 26]
```diff
@@ -0,0 +1,769 @@
...
+                _ = stop.changed() => true,
+                _ = check_interval.tick() => false,
+            };
+            if stop {
+                return Ok(true);
+            }
+            let result = client
+                .root_manager
+                .client()
+                .get_transaction_receipt(*pending_hash)
```
Comment (limemloh) [inline relayer/src/merkle.rs:608]: Can't really find any useful documentation for `get_transaction_receipt`, but if it somehow ends up in a state where this keeps returning None, (like the transaction is lost or something) this will loop forever.
Not sure if this is realistic though

File: relayer/src/merkle.rs [hunk 27]
```diff
@@ -0,0 +1,769 @@
...
+                return Ok(());
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Wait until the transaction is no longer pending.
+/// Return whether the process should be stopped.
+async fn wait_pending_ethereum_tx<M: Middleware, S: Signer>(
```
Comment (limemloh) [inline relayer/src/merkle.rs:583]: This function is hard to follow, it seems to me that there are essentially two loops going, one loop for waiting for the transaction to enter a block, another loop for waiting until this block is "finalized".
The rest of the logic seems to be "outside" of the loop.
If this is the case, then consider structuring the code like this.

File: relayer/src/merkle.rs (hunk 21 above)
Comment (abizjak) [reply relayer/src/merkle.rs:689]: Yes, it used to be needed when the types were different. It's a leftover from there.

File: relayer/src/merkle.rs (hunk 22 above)
Comment (abizjak) [reply relayer/src/merkle.rs:699]: Same here.

File: relayer/src/merkle.rs (hunk 23 above)
Comment (abizjak) [reply relayer/src/merkle.rs:594]: I don't think this helps, so I won't do it.

File: relayer/src/merkle.rs (hunk 24 above)
Comment (abizjak) [reply relayer/src/merkle.rs:612]: I have revised this using `let else`.

File: relayer/src/merkle.rs (hunk 25 above)
Comment (abizjak) [reply relayer/src/merkle.rs]: I don't think will improve things, so I am not doing it.

File: relayer/src/merkle.rs (hunk 26 above)
Comment (abizjak) [reply relayer/src/merkle.rs:608]: Yes, that is possible, and one of the remaining cases where we need to figure out what to do.

The same goes for Concordium transactions, there we also don't handle the case where they go missing. At least with Concordium we know the scenarios wehre that can happen. With Ethereum we get no guarantees from anybody, so it's a bit hard to know what to do. Restarting will help though.

File: relayer/src/merkle.rs (hunk 27 above)
Comment (abizjak) [reply relayer/src/merkle.rs:583]: I am not sure what exactly you are suggesting.

We are waiting, and once the waiting is done we do the action. I could move the action to a separate function, but I don't think that will help understanding. I can add documentation.

And it's not like we can get the transaction receipt and then wait for enough confirmations. We have to keep re-requesting everything due to the amazing guarantees we get.

Comment (limemloh) [reply relayer/src/merkle.rs:583]: okay, I see. Then it have to be one loop.

**Review by limemloh — COMMENTED** [review]
I don't have any more comments on the relayer, other than nitpicking. 
Maybe use `.context(..)` a bit more, just to improve the error messages.

File: relayer/src/merkle.rs (hunk 26 above)
Comment (abizjak) [reply relayer/src/merkle.rs:654]: This is now handled and tested as much as I could (by setting too low gas price to start with).
